File: examples/go/batch_csv.go
package main

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/joho/godotenv"
)

const (
	defaultActorID = "apple_yang/instagram-transcripts-scraper"
	apiBaseURL     = "https://api.apify.com/v2"
	maxRetries     = 2
)

type apiResponse struct {
	Data runInfo `json:"data"`
}

type runInfo struct {
	ID               string `json:"id"`
	Status           string `json:"status"`
	DefaultDatasetID string `json:"defaultDatasetId"`
}

type batchResult struct {
	SourceURL string           `json:"sourceUrl"`
	Status    string           `json:"status"`
	ItemCount int              `json:"itemCount"`
	Text      string           `json:"text"`
	Segments  []any            `json:"segments"`
	ErrMsg    string           `json:"errMsg"`
	RawItems  []map[string]any `json:"rawItems"`
}

type batchError struct {
	SourceURL string
	Error     string
}

func main() {
	if err := runBatch(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}

func runBatch() error {
	repoRoot, err := findRepoRoot()
	if err != nil {
		return err
	}
	_ = godotenv.Load(filepath.Join(repoRoot, ".env"))

	token, err := requireEnv("APIFY_TOKEN")
	if err != nil {
		return err
	}
	actorID := envOrDefault("APIFY_ACTOR_ID", defaultActorID)
	sessionID := os.Getenv("INSTAGRAM_SESSIONID")
	maxConcurrency, err := parseMaxConcurrency(envOrDefault("MAX_CONCURRENCY", "3"))
	if err != nil {
		return err
	}
	csvPath := filepath.Join(repoRoot, "sample-data", "instagram-reel-urls.csv")
	if len(os.Args) > 1 {
		csvPath = os.Args[1]
		if !filepath.IsAbs(csvPath) {
			abs, err := filepath.Abs(csvPath)
			if err != nil {
				return err
			}
			csvPath = abs
		}
	}

	urls, err := readURLs(csvPath)
	if err != nil {
		return err
	}
	if len(urls) == 0 {
		return errors.New("Error: CSV does not contain any videoUrl values.")
	}

	fmt.Println("Actor ID:", actorID)
	fmt.Println("Total URLs:", len(urls))
	fmt.Println("Max concurrency:", maxConcurrency)
	// Higher concurrency may increase Apify usage cost and rate-limit risk.

	client := &http.Client{Timeout: 10 * time.Minute}
	jobs := make(chan string)
	results := make([]batchResult, 0)
	failures := make([]batchError, 0)
	var mu sync.Mutex
	var processed int64
	var success int64
	var failed int64
	var wg sync.WaitGroup

	for worker := 0; worker < maxConcurrency; worker++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for videoURL := range jobs {
				result, err := processURL(client, token, actorID, videoURL, sessionID)
				mu.Lock()
				if err != nil {
					failures = append(failures, batchError{SourceURL: videoURL, Error: err.Error()})
					atomic.AddInt64(&failed, 1)
				} else {
					results = append(results, result)
					atomic.AddInt64(&success, 1)
				}
				done := atomic.AddInt64(&processed, 1)
				fmt.Printf("Progress: total=%d processed=%d success=%d failed=%d\n", len(urls), done, atomic.LoadInt64(&success), atomic.LoadInt64(&failed))
				mu.Unlock()
			}
		}()
	}

	for _, videoURL := range urls {
		jobs <- videoURL
	}
	close(jobs)
	wg.Wait()

	outputDir := filepath.Join(repoRoot, "sample-output", "generated")
	if err := os.MkdirAll(outputDir, 0o755); err != nil {
		return err
	}
	resultsPath := filepath.Join(outputDir, "go-batch-results.jsonl")
	errorsPath := filepath.Join(outputDir, "go-batch-errors.csv")
	if err := writeJSONL(resultsPath, results); err != nil {
		return err
	}
	if err := writeErrorsCSV(errorsPath, failures); err != nil {
		return err
	}

	fmt.Println("Results path:", resultsPath)
	fmt.Println("Errors path:", errorsPath)
	return nil
}

func processURL(client *http.Client, token, actorID, videoURL, sessionID string) (batchResult, error) {
	var lastErr error
	for attempt := 0; attempt <= maxRetries; attempt++ {
		items, _, err := callActorOnce(client, token, actorID, videoURL, sessionID)
		if err == nil {
			return buildResult(videoURL, items), nil
		}
		lastErr = err
		if attempt < maxRetries {
			time.Sleep(time.Duration(1<<attempt) * time.Second)
		}
	}
	return batchResult{}, lastErr
}

func callActorOnce(client *http.Client, token, actorID, videoURL, sessionID string) ([]map[string]any, []byte, error) {
	run, err := startActorRun(client, token, actorID, videoURL, sessionID)
	if err != nil {
		return nil, nil, err
	}
	if run.ID == "" {
		return nil, nil, errors.New("Actor run response did not include an id.")
	}
	finishedRun, err := waitForRun(client, token, run.ID)
	if err != nil {
		return nil, nil, err
	}
	if finishedRun.Status != "SUCCEEDED" {
		return nil, nil, fmt.Errorf("Actor run finished with status %s.", finishedRun.Status)
	}
	if finishedRun.DefaultDatasetID == "" {
		return nil, nil, errors.New("Actor run completed without a defaultDatasetId.")
	}
	return fetchDatasetItems(client, token, finishedRun.DefaultDatasetID)
}

func buildResult(sourceURL string, items []map[string]any) batchResult {
	text := firstString(items, "text")
	segments := firstSlice(items, "segments")
	errMsg := firstString(items, "errMsg")
	status := "success"
	if errMsg != "" {
		status = "actor_error"
	} else if len(items) == 0 {
		status = "empty"
	}
	return batchResult{
		SourceURL: sourceURL,
		Status:    status,
		ItemCount: len(items),
		Text:      text,
		Segments:  segments,
		ErrMsg:    errMsg,
		RawItems:  items,
	}
}

func startActorRun(client *http.Client, token, actorID, reelURL, sessionID string) (runInfo, error) {
	body, err := json.Marshal(map[string]string{
		"videoUrl":  reelURL,
		"sessionid": sessionID,
	})
	if err != nil {
		return runInfo{}, err
	}
	actorIDForURL := strings.ReplaceAll(actorID, "/", "~")
	url := fmt.Sprintf("%s/acts/%s/runs", apiBaseURL, actorIDForURL)

	var response apiResponse
	if err := doJSON(client, token, http.MethodPost, url, body, &response); err != nil {
		return runInfo{}, err
	}
	return response.Data, nil
}

func waitForRun(client *http.Client, token, runID string) (runInfo, error) {
	for attempt := 0; attempt < 150; attempt++ {
		url := fmt.Sprintf("%s/actor-runs/%s", apiBaseURL, runID)
		var response apiResponse
		if err := doJSON(client, token, http.MethodGet, url, nil, &response); err != nil {
			return runInfo{}, err
		}
		switch response.Data.Status {
		case "SUCCEEDED", "FAILED", "TIMED-OUT", "ABORTED":
			return response.Data, nil
		}
		time.Sleep(2 * time.Second)
	}
	return runInfo{}, errors.New("Timed out waiting for Actor run to finish.")
}

func fetchDatasetItems(client *http.Client, token, datasetID string) ([]map[string]any, []byte, error) {
	url := fmt.Sprintf("%s/datasets/%s/items", apiBaseURL, datasetID)
	raw, err := doRaw(client, token, http.MethodGet, url, nil)
	if err != nil {
		return nil, nil, err
	}
	var items []map[string]any
	if len(bytes.TrimSpace(raw)) > 0 {
		if err := json.Unmarshal(raw, &items); err != nil {
			return nil, nil, err
		}
	}
	if items == nil {
		items = []map[string]any{}
	}
	pretty, err := json.MarshalIndent(items, "", "  ")
	if err != nil {
		return nil, nil, err
	}
	return items, pretty, nil
}

func doJSON(client *http.Client, token, method, url string, body []byte, target any) error {
	raw, err := doRaw(client, token, method, url, body)
	if err != nil {
		return err
	}
	if len(bytes.TrimSpace(raw)) == 0 {
		return nil
	}
	return json.Unmarshal(raw, target)
}

func doRaw(client *http.Client, token, method, url string, body []byte) ([]byte, error) {
	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	request, err := http.NewRequest(method, url, reader)
	if err != nil {
		return nil, err
	}
	request.Header.Set("Authorization", "Bearer "+token)
	request.Header.Set("Content-Type", "application/json")

	response, err := client.Do(request)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	raw, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, err
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d from Apify API: %s", response.StatusCode, string(raw))
	}
	return raw, nil
}

func readURLs(csvPath string) ([]string, error) {
	file, err := os.Open(csvPath)
	if err != nil {
		return nil, fmt.Errorf("Error: CSV file not found: %s", csvPath)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}
	if len(records) == 0 {
		return nil, errors.New("Error: CSV is empty.")
	}
	videoURLIndex := -1
	for index, header := range records[0] {
		if strings.TrimSpace(header) == "videoUrl" {
			videoURLIndex = index
			break
		}
	}
	if videoURLIndex == -1 {
		return nil, errors.New("Error: CSV must include a videoUrl column.")
	}

	urls := make([]string, 0, len(records)-1)
	for _, record := range records[1:] {
		if videoURLIndex < len(record) {
			value := strings.TrimSpace(record[videoURLIndex])
			if value != "" {
				urls = append(urls, value)
			}
		}
	}
	return urls, nil
}

func writeJSONL(path string, results []batchResult) error {
	var builder strings.Builder
	for _, result := range results {
		line, err := json.Marshal(result)
		if err != nil {
			return err
		}
		builder.Write(line)
		builder.WriteByte('\n')
	}
	return os.WriteFile(path, []byte(builder.String()), 0o644)
}

func writeErrorsCSV(path string, failures []batchError) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	if err := writer.Write([]string{"sourceUrl", "error"}); err != nil {
		return err
	}
	for _, failure := range failures {
		if err := writer.Write([]string{failure.SourceURL, failure.Error}); err != nil {
			return err
		}
	}
	return writer.Error()
}

func firstString(items []map[string]any, field string) string {
	for _, item := range items {
		if value, ok := item[field].(string); ok && value != "" {
			return value
		}
	}
	return ""
}

func firstSlice(items []map[string]any, field string) []any {
	for _, item := range items {
		if value, ok := item[field].([]any); ok {
			return value
		}
	}
	return []any{}
}

func requireEnv(name string) (string, error) {
	value := strings.TrimSpace(os.Getenv(name))
	if value == "" {
		return "", fmt.Errorf("Error: %s is required. Set it in .env or your environment.", name)
	}
	return value, nil
}

func envOrDefault(name, defaultValue string) string {
	value := strings.TrimSpace(os.Getenv(name))
	if value == "" {
		return defaultValue
	}
	return value
}

func parseMaxConcurrency(rawValue string) (int, error) {
	value, err := strconv.Atoi(rawValue)
	if err != nil {
		return 0, errors.New("Error: MAX_CONCURRENCY must be an integer.")
	}
	if value < 1 {
		return 0, errors.New("Error: MAX_CONCURRENCY must be at least 1.")
	}
	return value, nil
}

func findRepoRoot() (string, error) {
	current, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		if fileExists(filepath.Join(current, ".env.example")) && fileExists(filepath.Join(current, "sample-data")) {
			return current, nil
		}
		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
	}
	return "", errors.New("Error: Could not locate repository root.")
}

func fileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

File: examples/rust/src/bin/batch_csv.rs
use anyhow::{anyhow, bail, Context, Result};
use csv::{Reader, Writer};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::env;
use std::fs::{self, File};
use std::io::{BufWriter, Write};
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use tokio::sync::{Mutex, Semaphore};
use tokio::time::{sleep, Duration};

const DEFAULT_ACTOR_ID: &str = "apple_yang/instagram-transcripts-scraper";
const API_BASE_URL: &str = "https://api.apify.com/v2";
const MAX_RETRIES: usize = 2;

#[derive(Debug, Deserialize)]
struct ApiResponse<T> {
    data: T,
}

#[derive(Debug, Deserialize)]
struct RunInfo {
    id: Option<String>,
    status: Option<String>,
    #[serde(rename = "defaultDatasetId")]
    default_dataset_id: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct BatchRecord {
    source_url: String,
    status: String,
    item_count: usize,
    text: String,
    segments: Vec<Value>,
    err_msg: String,
    raw_items: Vec<Value>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct ErrorRecord {
    source_url: String,
    error: String,
}

#[derive(Clone)]
struct Config {
    token: String,
    actor_id: String,
    sessionid: String,
}

#[tokio::main]
async fn main() -> Result<()> {
    let repo_root = find_repo_root()?;
    let _ = dotenvy::from_path(repo_root.join(".env"));

    let config = Config {
        token: require_env("APIFY_TOKEN")?,
        actor_id: env::var("APIFY_ACTOR_ID").unwrap_or_else(|_| DEFAULT_ACTOR_ID.to_string()),
        sessionid: env::var("INSTAGRAM_SESSIONID").unwrap_or_default(),
    };
    let max_concurrency = parse_max_concurrency()?;
    let csv_path = env::args()
        .nth(1)
        .map(PathBuf::from)
        .unwrap_or_else(|| repo_root.join("sample-data/instagram-reel-urls.csv"));
    let urls = read_urls(&csv_path)?;
    if urls.is_empty() {
        bail!("Error: CSV does not contain any videoUrl values.");
    }

    println!("Actor ID: {}", config.actor_id);
    println!("Total URLs: {}", urls.len());
    println!("Max concurrency: {max_concurrency}");
    // Higher concurrency may increase Apify usage cost and rate-limit risk.

    let output_dir = repo_root.join("sample-output/generated");
    fs::create_dir_all(&output_dir).context("Failed to create output directory")?;

    let client = Client::builder()
        .timeout(Duration::from_secs(600))
        .build()
        .context("Failed to create HTTP client")?;
    let semaphore = Arc::new(Semaphore::new(max_concurrency));
    let results = Arc::new(Mutex::new(Vec::<BatchRecord>::new()));
    let errors = Arc::new(Mutex::new(Vec::<ErrorRecord>::new()));
    let processed = Arc::new(AtomicUsize::new(0));
    let success = Arc::new(AtomicUsize::new(0));
    let failed = Arc::new(AtomicUsize::new(0));
    let total = urls.len();
    let mut handles = Vec::with_capacity(urls.len());

    for url in urls {
        let permit = semaphore.clone().acquire_owned().await?;
        let client = client.clone();
        let config = config.clone();
        let results = results.clone();
        let errors = errors.clone();
        let processed = processed.clone();
        let success = success.clone();
        let failed = failed.clone();

        handles.push(tokio::spawn(async move {
            let _permit = permit;
            match process_url(&client, &config, &url).await {
                Ok(record) => {
                    results.lock().await.push(record);
                    success.fetch_add(1, Ordering::Relaxed);
                }
                Err(error) => {
                    errors.lock().await.push(ErrorRecord {
                        source_url: url,
                        error: error.to_string(),
                    });
                    failed.fetch_add(1, Ordering::Relaxed);
                }
            }
            let done = processed.fetch_add(1, Ordering::Relaxed) + 1;
            println!(
                "Progress: total={} processed={} success={} failed={}",
                total,
                done,
                success.load(Ordering::Relaxed),
                failed.load(Ordering::Relaxed)
            );
        }));
    }

    for handle in handles {
        handle.await.context("Batch worker failed")?;
    }

    let results_path = output_dir.join("rust-batch-results.jsonl");
    let errors_path = output_dir.join("rust-batch-errors.csv");
    write_jsonl(&results_path, &results.lock().await)?;
    write_errors_csv(&errors_path, &errors.lock().await)?;

    println!("Results path: {}", results_path.display());
    println!("Errors path: {}", errors_path.display());

    Ok(())
}

async fn process_url(client: &Client, config: &Config, url: &str) -> Result<BatchRecord> {
    let mut last_error = None;
    for attempt in 0..=MAX_RETRIES {
        match call_actor_once(client, config, url).await {
            Ok(items) => return Ok(build_record(url, items)),
            Err(error) => {
                last_error = Some(error);
                if attempt < MAX_RETRIES {
                    sleep(Duration::from_secs(2_u64.pow(attempt as u32))).await;
                }
            }
        }
    }
    Err(last_error.unwrap_or_else(|| anyhow!("Unknown error")))
}

async fn call_actor_once(client: &Client, config: &Config, reel_url: &str) -> Result<Vec<Value>> {
    let run =
        start_actor_run(client, &config.token, &config.actor_id, reel_url, &config.sessionid)
            .await?;
    let run_id = run
        .id
        .ok_or_else(|| anyhow!("Actor run response did not include an id."))?;
    let finished_run = wait_for_run(client, &config.token, &run_id).await?;
    let status = finished_run.status.unwrap_or_default();
    if status != "SUCCEEDED" {
        bail!("Actor run finished with status {status}.");
    }
    let dataset_id = finished_run
        .default_dataset_id
        .ok_or_else(|| anyhow!("Actor run completed without a defaultDatasetId."))?;
    fetch_dataset_items(client, &config.token, &dataset_id).await
}

fn build_record(source_url: &str, items: Vec<Value>) -> BatchRecord {
    let text = first_string(&items, "text").unwrap_or_default();
    let segments = first_array(&items, "segments");
    let err_msg = first_string(&items, "errMsg").unwrap_or_default();
    let status = if !err_msg.is_empty() {
        "actor_error"
    } else if items.is_empty() {
        "empty"
    } else {
        "success"
    };

    BatchRecord {
        source_url: source_url.to_string(),
        status: status.to_string(),
        item_count: items.len(),
        text,
        segments,
        err_msg,
        raw_items: items,
    }
}

async fn start_actor_run(
    client: &Client,
    token: &str,
    actor_id: &str,
    reel_url: &str,
    sessionid: &str,
) -> Result<RunInfo> {
    let actor_id_for_url = actor_id.replace('/', "~");
    let url = format!("{API_BASE_URL}/acts/{actor_id_for_url}/runs");
    let body = json!({
        "videoUrl": reel_url,
        "sessionid": sessionid,
    });

    let response: ApiResponse<RunInfo> =
        send_json(client, token, reqwest::Method::POST, &url, Some(body)).await?;
    Ok(response.data)
}

async fn wait_for_run(client: &Client, token: &str, run_id: &str) -> Result<RunInfo> {
    for _ in 0..150 {
        let url = format!("{API_BASE_URL}/actor-runs/{run_id}");
        let response: ApiResponse<RunInfo> =
            send_json(client, token, reqwest::Method::GET, &url, None).await?;
        let status = response.data.status.as_deref().unwrap_or("");
        if matches!(status, "SUCCEEDED" | "FAILED" | "TIMED-OUT" | "ABORTED") {
            return Ok(response.data);
        }
        sleep(Duration::from_secs(2)).await;
    }
    bail!("Timed out waiting for Actor run to finish.")
}

async fn fetch_dataset_items(client: &Client, token: &str, dataset_id: &str) -> Result<Vec<Value>> {
    let url = format!("{API_BASE_URL}/datasets/{dataset_id}/items");
    send_json(client, token, reqwest::Method::GET, &url, None).await
}

async fn send_json<T: for<'de> Deserialize<'de>>(
    client: &Client,
    token: &str,
    method: reqwest::Method,
    url: &str,
    body: Option<Value>,
) -> Result<T> {
    let mut request = client
        .request(method, url)
        .bearer_auth(token)
        .header("Content-Type", "application/json");
    if let Some(body) = body {
        request = request.json(&body);
    }

    let response = request.send().await.context("Apify API request failed")?;
    let status = response.status();
    let text = response.text().await.context("Failed to read Apify API response")?;
    if !status.is_success() {
        bail!("HTTP {} from Apify API: {}", status.as_u16(), text);
    }
    serde_json::from_str(&text).context("Failed to parse Apify API response JSON")
}

fn read_urls(csv_path: &Path) -> Result<Vec<String>> {
    let mut reader = Reader::from_path(csv_path)
        .with_context(|| format!("Error: CSV file not found: {}", csv_path.display()))?;
    let headers = reader.headers()?.clone();
    let video_url_index = headers
        .iter()
        .position(|header| header == "videoUrl")
        .ok_or_else(|| anyhow!("Error: CSV must include a videoUrl column."))?;

    let mut urls = Vec::new();
    for record in reader.records() {
        let record = record?;
        if let Some(value) = record.get(video_url_index) {
            let trimmed = value.trim();
            if !trimmed.is_empty() {
                urls.push(trimmed.to_string());
            }
        }
    }
    Ok(urls)
}

fn write_jsonl(path: &Path, records: &[BatchRecord]) -> Result<()> {
    let file = File::create(path).context("Failed to create JSONL output file")?;
    let mut writer = BufWriter::new(file);
    for record in records {
        writeln!(writer, "{}", serde_json::to_string(record)?)?;
    }
    Ok(())
}

fn write_errors_csv(path: &Path, records: &[ErrorRecord]) -> Result<()> {
    let mut writer = Writer::from_path(path).context("Failed to create errors CSV")?;
    writer.write_record(["sourceUrl", "error"])?;
    for record in records {
        writer.serialize(record)?;
    }
    writer.flush()?;
    Ok(())
}

fn first_string(items: &[Value], field: &str) -> Option<String> {
    items
        .iter()
        .filter_map(|item| item.get(field)?.as_str())
        .find(|value| !value.is_empty())
        .map(ToString::to_string)
}

fn first_array(items: &[Value], field: &str) -> Vec<Value> {
    items
        .iter()
        .filter_map(|item| item.get(field)?.as_array())
        .next()
        .cloned()
        .unwrap_or_default()
}

fn require_env(name: &str) -> Result<String> {
    let value = env::var(name).unwrap_or_default();
    if value.trim().is_empty() {
        bail!("Error: {name} is required. Set it in .env or your environment.");
    }
    Ok(value)
}

fn parse_max_concurrency() -> Result<usize> {
    let raw = env::var("MAX_CONCURRENCY").unwrap_or_else(|_| "3".to_string());
    let value = raw
        .parse::<usize>()
        .context("Error: MAX_CONCURRENCY must be an integer.")?;
    if value == 0 {
        bail!("Error: MAX_CONCURRENCY must be at least 1.");
    }
    Ok(value)
}

fn find_repo_root() -> Result<PathBuf> {
    let mut current = env::current_dir().context("Failed to read current directory")?;
    loop {
        if current.join(".env.example").exists() && current.join("sample-data").exists() {
            return Ok(current);
        }
        if !current.pop() {
            break;
        }
    }
    bail!("Error: Could not locate repository root.")
}
